etl/src/main/java/org/pih/hivmigration/etl/sql/VisitMigrator.groovy

```diff
@@ -0,0 +1,42 @@
+package org.pih.hivmigration.etl.sql
+
+class VisitMigrator extends SqlMigrator {
+
+    void migrate() {
+
+        executeMysql("Create visits from existing encounters", '''
+            -- Get visit type "Clinic or Hospital Visit"
+            SET @visit_type_id = (SELECT visit_type_id FROM visit_type WHERE uuid = 'f01c54cb-2225-471a-9cd5-d348552c337c');
+            SET @encounter_type_exclusions = '';  -- a comma-separated string like '1,2,3'
+            
+            INSERT INTO visit
+            (patient_id,   visit_type_id,  date_started,   date_stopped,   location_id,   creator,   date_created, voided, uuid)
+            SELECT  e.patient_id, @visit_type_id, e.date_started, e.date_stopped, e.location_id, e.creator, now(),     0, uuid()
+            FROM
+                (
+                    SELECT patient_id,
+                           Min(encounter_datetime) AS date_started,  -- encounter_datetime is always midnight
+                           Addtime(Max(encounter_datetime), '23:59:00') AS date_stopped,
+                           Max(location_id) as location_id,  /* Avoid using 'Unknown Location' if there is another location available */
+                           creator
+                    FROM   encounter
+                    WHERE  FIND_IN_SET(encounter_type, @encounter_type_exclusions) = 0
+                    GROUP  BY patient_id,
+                              Date(encounter_datetime)
+                ) AS e;
+            
+            -- Add visit IDs to their encounters
+            UPDATE encounter e
+                INNER JOIN visit v
+                ON e.patient_id = v.patient_id
+                    AND Date(e.encounter_datetime) = Date(v.date_started)
+            SET    e.visit_id = v.visit_id
+            WHERE FIND_IN_SET(e.encounter_type, @encounter_type_exclusions) = 0;
```

Comment:
I don't think this extra check on encounter type here is needed.

Reply:
Yeah, this is some example code that Brandon found on the OpenMRS wiki... doesn't do anything now, but would if we set up the encounter type exclusions above.

Reply:
My point was that we don't need this here since we are already doing it earlier above when we create visits.  By adding it here I think it is not needed because a) since we are migrating into an empty DB, the only visits that will exist are the one's we've already created, and we don't need to restrict further, and b) presumably we want to link any encounter to an existing visit if it's patient, date, and location match, even if it isn't in these encounter types...

Reply:
I think the idea here is that there may be some encounter types that never should be associated with visits... which is likely irrelevant in our case, but comes from the example code.

Reply:
I think there are encounter types that should not, on their own, create a visit or imply the patient came that day.  But if a visit exists already that matches the encounter date and patient and location, we could consider associating  it with it.  Not a big deal I guess.